### PR TITLE
Fused GEMM

### PR DESCRIPTION
**Pull request by jagrit06**
## Proposed changes

Fuses `gemm`, `addmm`, and `block_sparse_mm` into one Uber-shader that uses metal function constants for specialization 
No performance regression has been seen on a M2 Ultra 

## Checklist

Put an `x` in the boxes that apply.

- [x] I have read the [CONTRIBUTING](https://github.com/ml-explore/mlx/blob/main/CONTRIBUTING.md) document
- [x] I have run `pre-commit run --all-files` to format my code / installed pre-commit prior to committing changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have updated the necessary documentation (if needed)

### CODE REVIEW

File: mlx/backend/metal/device.cpp [hunk 1]
```diff
@@ -357,7 +357,7 @@ MTL::Function* Device::get_function_(
   }
 
   mtl_func_consts->release();
-  desc->release();
+  // desc->release();
```
Comment (awni) [inline mlx/backend/metal/device.cpp:360]: Should we delete that?

File: mlx/backend/metal/device.cpp [hunk 2]
```diff
@@ -526,11 +526,13 @@ MTL::ComputePipelineState* Device::get_kernel(
   // Compile kernel to compute pipeline
   auto mtl_linked_funcs = get_linked_functions_(linked_functions);
   auto kernel = get_kernel_(kname, mtl_function, mtl_linked_funcs);
-  mtl_function->release();
-  mtl_linked_funcs->release();
+
+  // mtl_function->release();
+  // mtl_linked_funcs->release();
```
Comment (awni) [inline mlx/backend/metal/device.cpp:531]: Are those not owned by MLX either? Maybe just delete as well?

File: mlx/backend/metal/matmul.cpp [hunk 3]
```diff
@@ -521,16 +549,46 @@ void steel_matmul(
...
+  // clang-format off
+  kname << "_has_batch_" << (has_batch ? 't' : 'n') 
+        << "_use_out_source_" << (use_out_source ? 't' : 'n') 
+        << "_do_axpby_" << (do_axpby ? 't' : 'n') 
+        << "_align_M_" << (align_M ? 't' : 'n')
+        << "_align_N_" << (align_N ? 't' : 'n') 
+        << "_align_K_" << (align_K ? 't' : 'n') 
+        << "_do_gather_" << (do_gather ? 't' : 'n'); // clang-format on
+
+  std::string hash_name = kname.str();
```
Comment (awni) [inline mlx/backend/metal/matmul.cpp:586]: Just curious, what is that used for?

**Review by awni — APPROVED** [review]
Looks great, thanks!! There are a couple stray comments related to releasing the objc objects that we can probably remove. Otherwise feel free to merge when ready!

File: mlx/backend/metal/matmul.cpp (hunk 3 above)
Comment (jagrit06) [reply mlx/backend/metal/matmul.cpp:586]: The do gather ? That does the `block_sparse_mm` variant

Comment (awni) [reply mlx/backend/metal/matmul.cpp:586]: I mean the `hash_name`

**Pull request merged**